libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java

```diff
@@ -1302,4 +1304,41 @@ private static void ensureNoSelfReferences(final Iterable<?> value, Object origi
         }
     }
 
+    /**
+     * Checks whether the given value is writeable as x-content.
+     *
+     * If the value cannot be passed to {@link #value(Object)} without
+     * error, an error {@link IllegalArgumentException} is thrown.
+     */
+    public static void ensureToXContentable(@Nullable Object value) {
+        if (value == null) {
+            return;
+        }
+
+        if (value instanceof Object[] array) {
+            for (Object v : array) {
+                ensureToXContentable(v);
+            }
+        } else if (value instanceof Map<?,?> map) {
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                ensureToXContentable(entry.getKey());
```

Comment:
Doesn't `unknownValue` require that the key is a `String`, and not just any writable value?

Reply:
Yep, added a check for String and a test.